pkg/yaml_util/yaml_util.go

```diff
@@ -0,0 +1,129 @@
+package yaml_util
+
+import (
+	"bytes"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v3"
+	"strings"
+)
+
+type CheckMode bool
+
+const (
+	Lenient = CheckMode(false)
+	Strict  = CheckMode(true)
+)
+
+// SetValue upserts the value the content yaml to a given value, specified by a dotted path. For example, setting
+// `foo.bar.baz` to `hello, world` is equivalent to upserting the following yaml:
+//
+//	foo:
+//	    bar:
+//	        baz: hello, world
+//
+// This method will make a best effort to preserve comments, as per the `yaml` package's abilities. You may overwrite
+// scalars, but you may not overwrite a non-scalar. You may also specify a path that doesn't exist in the source yaml,
+// as long as none of the paths correspond to existing elements other than yaml mappings.
+func SetValue(content []byte, optionPath string, optionValue string) ([]byte, error) {
+	// General approach:
+	// 1) convert the yaml into a map, using yaml
+	// 2a) find the node tree's value at the specified path
+	// 2b) set that node's value, assuming it's a scalar (or empty)
+	// 2) write the node tree back into bytes. this will preserve comments and such
+
+	// step 1
+	var tree yaml.Node
+	if err := yaml.Unmarshal(content, &tree); err != nil {
+		return nil, err
+	}
+
+	// step 2a
+	segments := strings.Split(optionPath, ".")
+	var topNode *yaml.Node
+	if len(tree.Content) == 0 {
+		topNode = &yaml.Node{Kind: yaml.MappingNode}
+	} else {
+		topNode = tree.Content[0] // the tree's root is a DocumentNode; we assume one document
+	}
+	setOptionAtNode := topNode
+	for _, segment := range segments[:len(segments)-1] {
```

Comment:
for 2a) are we finding the value at the non leaf path? ie up to bar in this scenario? I assume thats the intention, just making sure

Reply:
Correct. We want the node right before the last segment, and then in 2b we set its value at that last segment.

So in `fizz.buzz.foo = 123`, the 2a loop gets you the `fizz.buzz` node, and then 2b sets `*buzzNode["foo"] => 123`